### PR TITLE
[pytorch] Improve process_group_agent() serialization speed

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29785 [pytorch] Improve process_group_agent() serialization speed**

TLDR: This change improves process_group's serialization speed:

```
  Serialize_Tensor64:     12.38us ->   1.99us  (~-84%)
  Deserialize_Tensor64:   33.89us ->   5.62us  (~-84%)
  Serialize_Tensor1M:    525.74us -> 285.43us  (~-45%)
  Deserialize_Tensor1M:  892.61us -> 273.68us  (~-70%)
```

After speaking with the jit team, we had consensus that torch::save()/load()
are somewhat high-overhead for RPC serialization, mostly intended for
persistent disk data.

(Particularly, for large tensors, 35% of the time is spent in CRC checking, even
with the fb-side changes to subsitute 40x faster SSE-accelerated crc checking;
Also, for small tensors, the zip container overhead is considerable, as is the
overhead of lexing/parsing an embedded text python program for each RPC).

The jit team encouraged us to use jit::pickler, with the WriteableTensorData
way of outputting result tensors (not the default side-tensor table, or
with pickling the actual tensors). This ends up just pickling some tensor
metadata, and giving us some tensor blobs that we can mindlessly
blit over the wire (they copy to cpu memory if needed).

There is yet no standardized container format for the pickled data
(there is jit::pickle_save() checked in, but but it's experimental,
no load function is yet provided), but they encouraged us to just use
something sensible for this, and possibly revisit later. For now, I made
the directory headers slightly http-inspired.

Note that serialization is just one component of the pipeline, but that
said, we also see reasonable reductions in end-to-end echo times (noisier):

```
   ProcessGroupAgent_Echo(Tensor_Small)   855.25us -> 492.65us  (~-42%)
   ProcessGroupAgent_Echo(Tensor_1M)       10.82ms -> 6.94ms    (~-35%)
   ProcessGroupAgent_Echo(Small_NoTensor) 688.82us -> 301.72us  (~-56%)
   ProcessGroupAgent_Echo(1MB_NoTensor)     4.65ms -> 3.71ms    (~-20%)
```

I moved the "wire serialization" logic to a separate file to assist with
unittesting.

Differential Revision: [D18493938](https://our.internmc.facebook.com/intern/diff/D18493938/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/utils.cpp [hunk 1]
```diff
@@ -101,6 +103,178 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+
+  if (!tensors.empty()) {
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          metaEntry.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensors);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/utils.cpp]: Question: how does `wireSerialize` compare to directly push everything into the jit pickler?

File: test/cpp/rpc/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+target_link_libraries(test_cpp_rpc PRIVATE torch gtest)
+target_include_directories(test_cpp_rpc PRIVATE ${ATen_CPU_INCLUDE})
+
+if (USE_CUDA)
+  target_link_libraries(test_cpp_rpc PRIVATE
+    ${CUDA_LIBRARIES}
+    ${CUDA_NVRTC_LIB}
+    ${CUDA_CUDA_LIB}
+    ${TORCH_CUDA_LIBRARIES})
```
Comment (pietern) [inline test/cpp/rpc/CMakeLists.txt]: Isn't there a `torch_cuda` or equivalent target that pulls all of this in?

**Review by pietern — COMMENTED** [review]
The new test directory doesn't seem to be used anywhere.

If this is not compiled, it will rot and become unusable over time. If you don't care about running it as part of the test suite, it should still be part of the default set of tests that get compiled.

Regarding the string format and generation/parsing: is all tensor data now inlined?

File: torch/csrc/distributed/rpc/utils.cpp (hunk 1 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: My understanding here is that the main viable alternative, if you need to send over the wire, is effectively torch::save(), which is several times slower, per the diff description.

If you provide a tensorTable, it will just push_back into the table (via pushTensorReference()), but not with metadata or anything useful - you're still stuck where you began with respect to getting these tensors in a blittable format.

Here, we're basically using the pickler with the tensors like ScriptModuleSerializer::writeArchive() does (see torch/csrc/jit/export.cpp) as part of torch::save(), so it's a fairly well-used codepath. The ScriptModuleSerializer use of this codepath then just puts the tensor files "0", "1", "2", etc in a zip archive, while we're doing something quite similar, but in a more streamlined manner.

The efficiency gains come chiefly from
  - avoiding CRC overhead (~35% of time, even with non-open-sourced 40x faster CRC function)
  - avoiding unnecessary copies and L2 miss that come from zip format
  - avoiding sending a per-call python program that takes ~15usec to Lex/Parse in every torch::load() call, even for a minimal tensor

File: test/cpp/rpc/CMakeLists.txt (hunk 2 above)
Comment (jjlilley) [reply test/cpp/rpc/CMakeLists.txt]: Pieter,

I was using TARGETS/fb side to test, and trying to sprinkle parallel CMakeLists.txt incantations to attempt to make things work (and did a few re-uploads last night where I thought I got things to work). I was going back-and-forth about putting in the existing cpp/api/serialize.cpp test, and think I might just move the code back there.

Re: string format- I want to make sure I understand your question, but yes, the tensor data is contained in the emitted string from that function (as it was in torch::save()

**Comment by jjlilley** [comment]
Sorry for the long set of uploads earlier.
I believe the CMakeLists is fixed up so that the wire serialization test run in the open source case.

Per my earlier comment to Shen, this is somewhat similar to the logic inside torch::save()'s call inside ScriptModuleSerializer::writeArchive(), but avoids much of the rest of the unneeded overhead.

**Comment by pietern** [comment]
@driazati @suo Could you chime in on this? Are there plans to pull this or similar functionality into the jit's pickler, so that we don't need a separate serialization mechanism? It's in everyone's interest if we limit the proliferation of picklers/serializers.

**Comment by jjlilley** [comment]
fwiw, some more context:
 - this is partly based on the FB version in caffe2/torch/fb/distributed/wireSerialization, and is being used in the Thrift RPC agent.
  - It seemed sensible to have two distinct-albeit-similar versions, since in the Thrift context, we wanted to have access to folly::IOBuf for efficiency

**Comment by driazati** [comment]
> @driazati @suo Could you chime in on this? Are there plans to pull this or similar functionality into the jit's pickler, so that we don't need a separate serialization mechanism? It's in everyone's interest if we limit the proliferation of picklers/serializers.

The pickler is narrowly focused on being a trimmed down implementation of Python's `pickle` format. On top of that is TorchScript's zip-file serialization (which contains several pickle files). Outside of that I don't think the TorchScript serialization should gain any functionality. In this case the zip-file has too much overhead for a wire format, so more direct access to the `jit::Pickler` is warranted.

File: torch/csrc/distributed/rpc/utils.cpp [hunk 3]
```diff
@@ -101,6 +103,178 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+    ++ptr; // past the '\n'
+  }
+  ++ptr; // past the final \n
+
+  std::unordered_map<std::string, std::pair<const char*, size_t>> out;
+  for (size_t i = 0; i < header_ents.size(); ++i) {
+    out[header_ents[i].first] = {ptr, header_ents[i].second};
+    ptr += header_ents[i].second;
+  }
+  if (!ok || ptr > endp) {
```
Comment (pietern) [inline torch/csrc/distributed/rpc/utils.cpp]: Use `ptr != endp` instead? Seems like it should always be true if the format was correctly followed.

**Review by pietern — COMMENTED** [review]
One minor error checking comment. Otherwise this LGTM. Very nice speedup :-)

File: torch/csrc/distributed/rpc/utils.cpp (hunk 3 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: Changed to equality checks per your comment, also fixed some clang-tidy issues.
Ready for another look, when you get a chance.

**Comment by jjlilley** [comment]
Checking: are there any other blockers for this?

**Review by pietern — COMMENTED** [review]
I don't think so, this seems good to go.

@mrshenli Any concerns?

File: torch/csrc/distributed/rpc/utils.cpp [hunk 4]
```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          metaEntry.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensors);
+    pickler.stop();
+    auto writeable_tensors = pickler.tensorData();
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/utils.cpp]: what prevents these tensors being destructed before we add them into `out`?

File: torch/csrc/distributed/rpc/utils.cpp [hunk 5]
```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+      if (meta_data_pos >= meta_data.second || n == 0) {
+        return 0;
+      }
+      size_t to_copy =
+          std::min(meta_data_pos + n, meta_data.second) - meta_data_pos;
+      memcpy(buf, meta_data.first + meta_data_pos, to_copy);
+      meta_data_pos += to_copy;
+      return to_copy;
+    };
+    auto section_read_func = [&](const std::string& fname) -> at::DataPtr {
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/utils.cpp]: What does this do? Why it needs a file name? Or is it actually Ent name?

**Review by mrshenli — COMMENTED** [review]
Thanks for adding this awesome perf optimization!! 

I have some questions commented inline.

File: torch/csrc/distributed/rpc/utils.cpp [hunk 6]
```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+  const char* endp = ptr + data_size;
+
+  std::vector<std::pair<std::string, size_t>> header_ents;
+  bool ok = false;
+  while (ptr != endp) {
+    if (*ptr == '\n') {
+      ok = true;
+      break;
+    }
+    const char* num_ptr = ptr;
```
[the remark this answers is not in the document]
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: You're right in that this should be "name," I'll change the varname and add a comment.
Also, Satendra reminded me of camelCase, so switched this as well.

File: torch/csrc/distributed/rpc/utils.cpp [hunk 7]
```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+      ptr++;
+    }
+    if (ptr == endp) {
+      break;
+    }
+    size_t sz = c10::stoll(std::string(size_ptr, ptr - size_ptr));
+    header_ents.emplace_back(std::make_pair(name, sz));
+    ++ptr; // past the '\n'
+  }
+  ++ptr; // past the final \n
```
[the remark this answers is not in the document]
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: Good point in that the code is more clear if this increment right next to the if (*ptr == '\n')  clause.

(correctness was same either way, due to 'ok' only being set in the case, but easier to reason about with your suggestion)

File: torch/csrc/distributed/rpc/utils.cpp [hunk 8]
```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
...
+
```
[the remark this answers is not in the document]
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: Changed to a separate pre-check for ok here.

Agree this makes the code more clear. I was simply trying to avoid an extra throw statement, which was probably slightly silly.

File: torch/csrc/distributed/rpc/utils.cpp (hunk 4 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: In torch/csrc/jit/pickler.h, the definition is as follows:

  const std::vector<WriteableTensorData>& tensorData() {                                                                                                  
    return tensor_data_;                                                                                                                                  
  }               

The tensorData() persisted by the pickler.

(Note the parallel usage in ScriptModuleSerializer::writeArchive() in torch/csrc/jit/export.cpp, as used by torch::save() - it's effectively the same usage idiom, but not put in a zip file).

File: torch/csrc/distributed/rpc/utils.cpp (hunk 5 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/utils.cpp]: It really is an entity name, I can change to that.

fwiw, in the context of torch::save() it kind of is a filename, since torch::save uses zip container format, with a directory full of these entities. But "entity" is probably more clear.

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Comment by jjlilley** [comment]
Thanks for reviewing!
I'll hold landing until Rohan gets in the pending 1.4 cherrypicks that he sent out.

**Comment by mrshenli** [comment]
The failures do not look relevant, but it starts to appear since this PR on master (checkout this link https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master). @jjlilley do you know what could cause this?

failures:
https://app.circleci.com/jobs/github/pytorch/pytorch/3738271
https://app.circleci.com/jobs/github/pytorch/pytorch/3738272

**Comment by jjlilley** [comment]
Thanks - saw this in the notifications yesterday.
Unclear to me so far why this is failing (test_fake_quant seems to work ok for me locally, and CircleCI was fully green when I landed), or why this would be even be vaguely related.

**Comment by suo** [comment]
@jjlilley Any progress on this? If not, I think we should revert this change before everyone returns from the holidays to find master broken :/

**Comment by jjlilley** [comment]
@suo That's sensible to do, since I don't see how this is related.
 I just wrote up a rollback for the (active) process_group_agent.{cpp,h} part at:
    https://github.com/pytorch/pytorch/pull/30593

**Comment by mrshenli** [comment]
@jjlilley any reason for adding a new PR instead of reverting this one?

**Comment by suo** [comment]
I'm interested to see if the tests are actually fixed by a revert, so let's watch the CI for the rollback PR.

**Comment by jjlilley** [comment]
@mrshenli no strong reason, I’m just more fluent in Phab‘s knobs

**Comment by jjlilley** [comment]
@suo Thanks for the preemptive accept on 30593.
That said, rolling back doesn't help (the failure seemed quite unrelated anyway)

Maybe the best way forward is to disable the test_backward_per_tensor testcase in test_fake_quant?
(I just sent out https://github.com/pytorch/pytorch/pull/30594 which does this)
